engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java

```diff
@@ -191,6 +199,51 @@ Text toDisplayString(boolean allowSideEffects, @CachedLibrary("this") InteropLib
     }
   }
 
+  @ExportMessage
+  int identityHashCode(
+      @CachedLibrary(limit = "5") InteropLibrary fieldsInterop,
+      @Cached(value = "computeIdentityHashCode(this, fieldsInterop)", allowUncached = true) int cachedHashCode
+  ) {
+    return cachedHashCode;
+  }
+
+  @ExportMessage
+  static class IsIdenticalOrUndefined {
+    @Specialization
+    static TriState isIdenticalOrUndefined(
+        Atom thisAtom,
+        Atom otherAtom,
+        @CachedLibrary(limit = "5") InteropLibrary fieldsInterop,
+        @Cached(value = "computeIdentityHashCode(thisAtom, fieldsInterop)", allowUncached = true) int thisHashCode,
+        @Cached(value = "computeIdentityHashCode(otherAtom, fieldsInterop)", allowUncached = true) int otherHashCode
+    ) {
+      return thisHashCode == otherHashCode ? TriState.TRUE : TriState.FALSE;
+    }
+
+    @Fallback
+    static TriState fallBack(Atom thisAtom, Object other) {
+      return TriState.FALSE;
+    }
+  }
+
+  @ExplodeLoop
+  static int computeIdentityHashCode(Atom atom, InteropLibrary fieldsInterop) {
+    int[] hashCodes = new int[atom.fields.length];
+    for (int i = 0; i < atom.fields.length; i++) {
+      Object field = atom.fields[i];
+      if (fieldsInterop.hasIdentity(field)) {
+        try {
+          hashCodes[i] = fieldsInterop.identityHashCode(field);
```

Comment:
Is `identityHashCode` the right thing to use? I always thought it is a Truffle replacement for `System.indentityHashCode` and that's not what one to use when computing `hashCode`. Can you point me to some examples in other languages that use/implement this message?

Reply:
> Is identityHashCode the right thing to use?

I am not sure anymore. But what other message from `InteropLibrary` could we use? But since we export `identityHashCode` message in `Atom`, and we also could export this message for other builtins, I guess this is OK. For primitives, `identityHashCode` and `hashCode` should be the same, if I am not mistaken.

Reply:
`identityHashCode` is, indeed, an equivalent to `System.identityHashCode`, which we do not want to use. I got rid of all the `identityHashCode` messages from all our classes.